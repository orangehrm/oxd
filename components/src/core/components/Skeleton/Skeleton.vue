<template>
  <div v-for="line in lines" :key="line" :class="classes" :style="styles"></div>
</template>

<script lang="ts">
import {defineComponent} from 'vue';

export default defineComponent({
  name: 'oxd-skeleton',

  props: {
    animate: {
      type: Boolean,
      default: false,
    },
    circle: {
      type: Boolean,
      default: false,
    },
    height: {
      type: [String, Number],
      default: null,
    },
    width: {
      type: [String, Number],
      default: null,
    },
    lines: {
      type: Number,
      default: 1,
    },
  },

  computed: {
    classes() {
      return {
        'oxd-skeleton': true,
        '--circle': this.circle,
        '--animate': this.animate,
      };
    },
    styles() {
      const width =
        typeof this.width === 'number' ? `${this.width}px` : this.width;
      const height =
        typeof this.height === 'number' ? `${this.height}px` : this.height;

      return {
        ...(width && {width}),
        ...(height && {height}),
      };
    },
  },
});
</script>

<style lang="scss" src="./skeleton.scss" scoped></style>
