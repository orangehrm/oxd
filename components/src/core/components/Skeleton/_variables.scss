$skeleton-background-color: $oxd-background-light-gray-color !default;
$skeleton-shine-color: $oxd-interface-gray-lighten-1-color !default;
