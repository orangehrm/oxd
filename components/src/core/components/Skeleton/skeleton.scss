@import '../../../styles';
@import 'variables';

.oxd-skeleton {
  width: 100%;
  height: 1.2rem;
  overflow: hidden;
  position: relative;
  margin-bottom: 5px;
  border-radius: $oxd-border-radius;
  background-color: $skeleton-background-color;

  &.--animate {
    &::after {
      content: '';
      position: absolute;
      top: 0;
      left: 0;
      z-index: 1;
      width: 100%;
      height: 100%;
      animation: loading 1.5s infinite;
      background: linear-gradient(
        90deg,
        rgba($skeleton-shine-color, 0),
        rgba($skeleton-shine-color, 0.3),
        rgba($skeleton-shine-color, 0)
      );
      transform: translate(-100%);
    }
  }

  &.--circle {
    width: 1.2rem;
    border-radius: 100%;
  }
}

@keyframes loading {
  to {
    transform: translate(100%);
  }
}
