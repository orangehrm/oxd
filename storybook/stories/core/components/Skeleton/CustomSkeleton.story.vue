<template>
  <oxd-grid :cols="4">
    <oxd-grid-item v-for="i in 12" :key="i">
      <oxd-skeleton :width="50" :height="50" circle animate />
      <oxd-skeleton :lines="3" animate />
    </oxd-grid-item>
  </oxd-grid>
</template>

<script>
import Grid from '@orangehrm/oxd/core/components/Grid/Grid';
import GridItem from '@orangehrm/oxd/core/components/Grid/GridItem';
import Skeleton from '@orangehrm/oxd/core/components/Skeleton/Skeleton.vue';

export default {
  components: {
    'oxd-grid': Grid,
    'oxd-skeleton': Skeleton,
    'oxd-grid-item': GridItem,
  },
};
</script>
