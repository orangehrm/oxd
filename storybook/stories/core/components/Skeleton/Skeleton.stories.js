import Skeleton from '@orangehrm/oxd/core/components/Skeleton/Skeleton.vue';
import CustomSkeletonStory from './CustomSkeleton.story.vue';

export default {
  title: 'Panels/Skeleton',
  component: Skeleton,
  argTypes: {
    animate: {
      control: {type: 'boolean'},
      table: {
        type: {summary: 'Whether to show animation'},
      },
    },
    circle: {
      control: {type: 'boolean'},
      table: {
        type: {summary: 'Whether to show as circle'},
      },
    },
    width: {
      control: {type: 'number'},
      table: {
        type: {summary: 'Set the width of skeleton element'},
      },
    },
    height: {
      control: {type: 'number'},
      table: {
        type: {summary: 'Set the height of skeleton element'},
      },
    },
    lines: {
      control: {type: 'number'},
      defaultValue: 1,
      table: {
        type: {summary: 'Set the number of repetitions'},
      },
    },
  },
};

const Template = (args) => ({
  setup() {
    return {args};
  },
  components: {'oxd-skeleton': Skeleton},
  template: '<oxd-skeleton v-bind="args" />',
});

export const Default = Template.bind({});
Default.args = {
  width: 100,
  height: 100,
};
Default.parameters = {
  docs: {
    source: {
      code: `
<oxd-skeleton :width="100" :height="100" />
    `,
    },
  },
};

export const Animate = Template.bind({});
Animate.args = {
  width: 100,
  height: 100,
  animate: true,
};
Animate.parameters = {
  docs: {
    source: {
      code: `
<oxd-skeleton :width="100" :height="100" :animate="true" />
    `,
    },
  },
};

export const Circle = Template.bind({});
Circle.args = {
  width: 100,
  height: 100,
  circle: true,
  animate: true,
};
Circle.parameters = {
  docs: {
    source: {
      code: `
<oxd-skeleton :width="100" :height="100" :circle="true" :animate="true" />
    `,
    },
  },
};

export const Custom = () => CustomSkeletonStory;
Custom.parameters = {
  docs: {
    source: {
      code: `
<oxd-grid :cols="4">
  <oxd-grid-item v-for="i in 12" :key="i">
    <oxd-skeleton :width="50" :height="50" circle animate />
    <oxd-skeleton :lines="3" animate />
  </oxd-grid-item>
</oxd-grid>
    `,
    },
  },
};
