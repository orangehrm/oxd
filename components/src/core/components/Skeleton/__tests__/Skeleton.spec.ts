import {mount} from '@vue/test-utils';
import Skeleton from '@orangehrm/oxd/core/components/Skeleton/Skeleton.vue';

describe('Skeleton.vue', () => {
  it('should renders OXD Skeleton', () => {
    const wrapper = mount(Skeleton);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should renders OXD Skeleton, 3 lines', () => {
    const wrapper = mount(Skeleton, {
      props: {lines: 3},
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should renders OXD Skeleton > Circle with custom height and width', () => {
    const wrapper = mount(Skeleton, {
      props: {
        circle: true,
        height: 100,
        width: 100,
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should renders OXD Skeleton > with Animation', () => {
    const wrapper = mount(Skeleton, {
      props: {
        animate: true,
        height: 100,
        width: 100,
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
