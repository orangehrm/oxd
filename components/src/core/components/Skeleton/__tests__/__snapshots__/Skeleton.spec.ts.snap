// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Skeleton.vue should renders OXD Skeleton > Circle with custom height and width 1`] = `<div class="oxd-skeleton --circle" style="width: 100px; height: 100px;"></div>`;

exports[`Skeleton.vue should renders OXD Skeleton > with Animation 1`] = `<div class="oxd-skeleton --animate" style="width: 100px; height: 100px;"></div>`;

exports[`Skeleton.vue should renders OXD Skeleton 1`] = `<div class="oxd-skeleton"></div>`;

exports[`Skeleton.vue should renders OXD Skeleton, 3 lines 1`] = `
<div class="oxd-skeleton"></div>
<div class="oxd-skeleton"></div>
<div class="oxd-skeleton"></div>
`;
